internal/transports/p2p/peer/checkpoint.go

```diff
@@ -0,0 +1,139 @@
+package peer
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/bitcoin-sv/block-headers-service/domains"
+	"github.com/bitcoin-sv/block-headers-service/internal/chaincfg"
+	"github.com/bitcoin-sv/block-headers-service/internal/chaincfg/chainhash"
+	"github.com/rs/zerolog"
+)
+
+type checkpoint struct {
+	checkpoint      *chaincfg.Checkpoint
+	currentIndex    int
+	finalCheckpoint *chaincfg.Checkpoint
+	checkpoints     []chaincfg.Checkpoint
+	log             *zerolog.Logger
+	lock            sync.RWMutex
+}
+
+func newCheckpoint(checkpoints []chaincfg.Checkpoint, tipHeight int32, log *zerolog.Logger) *checkpoint {
+	logger := log.With().Str("subservice", "checkpoint").Logger()
+	ch := &checkpoint{
+		checkpoints: checkpoints,
+		log:         &logger,
+	}
+
+	if len(checkpoints) != 0 {
+		ch.finalCheckpoint = &checkpoints[len(checkpoints)-1]
+		ch.next(tipHeight)
+	}
+
+	return ch
+}
+
+func (ch *checkpoint) Height() int32 {
+	ch.lock.RLock()
+	defer ch.lock.RUnlock()
+
+	return ch.checkpoint.Height
+}
+
+func (ch *checkpoint) Hash() *chainhash.Hash {
+	ch.lock.RLock()
+	defer ch.lock.RUnlock()
+
+	return ch.checkpoint.Hash
+}
+
+// LastReached returns true if the last checkpoint has been reached.
+func (ch *checkpoint) LastReached() bool {
+	ch.lock.RLock()
+	defer ch.lock.RUnlock()
+
+	return ch.checkpoint == nil
+}
+
+// Check checks if the header is valid according to the checkpoint and marks switches to next checkpoint if reached.
+func (ch *checkpoint) Check(header *domains.BlockHeader) error {
```

Comment:
A minor, but I like the `Verify` name much more than just `Check` for this method 😅 

Or maybe even `VerifyAndUpdate` because it's also updating to the next checkpoint? 🤔

Reply:
I'm not convinced to that
1. `checkpoint` seem to be good source for `checking` not verifying because it's not `verificationpoint` 😄 
2. I was thinking about being more clear method name, but it's hard decision because from the readability and abstraction point of view, `Check` seems more proper - it shouldn't bother you how it is implemented below, if it is storing the checkpoint or if it is searching for it each time. Maybe the only thing that should be kinda published is the error that we received an unexpected height that is above the last checkpoint 🤷 

To wrap up:
1. I'm rather voteing for `Check` instead of `Verify`, but if someone else would vote for verify, I can live with that (and rename it)
2. I'm not convinced if we really need to add `Update` or other suffix telling about side effects, so the third vote would be necessary
FYI @chris-4chain ☝️

Reply:
> checkpoint seem to be good source for checking not verifying because it's not verificationpoint 😄

⬆️  It's funny but it is true 🙈 

About point `2.`. @kuba-4chain has a point that the name: `Check` suggests that it is a "const method".
> In C++ const methods `int function_name () const` are denied to change the state of the instance. 

In this case the `Check` methods modifies it...
The word "check" often implies a passive operation...
🤔 
🤔 
My types:
`CheckAndAdvance`, `Process`, `CheckAndProgress`, `Handle`

Reply:
> checkpoint seem to be good source for checking not verifying because it's not verificationpoint 😄

> ⬆️ It's funny but it is true 🙈

But we're not exactly `checking` the `point`. What we're doing is `verifying` whether the checkpoint hash is equal to the block hash we received. We're actually `verifying` the block header against the checkpoint.